common/hsm_secret.c

```diff
@@ -0,0 +1,463 @@
+				      const char *passphrase,
+				      enum hsm_secret_error *err)
+{
+	enum hsm_secret_type type = detect_hsm_secret_type(hsm_secret, len);
+	
+	switch (type) {
+	case HSM_SECRET_PLAIN:
+		return extract_plain_secret(ctx, hsm_secret, len, err);
+	case HSM_SECRET_ENCRYPTED:
+		return extract_encrypted_secret(ctx, hsm_secret, len, passphrase, err);
+	case HSM_SECRET_MNEMONIC_NO_PASS:
+	case HSM_SECRET_MNEMONIC_WITH_PASS:
+		return extract_mnemonic_secret(ctx, hsm_secret, len, passphrase, type, err);
+	case HSM_SECRET_INVALID:
+		*err = HSM_SECRET_ERR_INVALID_FORMAT;
+		return NULL;
+	}
+	abort();
+}
+
+bool encrypt_legacy_hsm_secret(const struct secret *encryption_key,
+			const struct secret *hsm_secret,
+			u8 *output)
+{
+	crypto_secretstream_xchacha20poly1305_state crypto_state;
+
+	if (crypto_secretstream_xchacha20poly1305_init_push(&crypto_state, output,
+							    encryption_key->data) != 0)
+		return false;
+	if (crypto_secretstream_xchacha20poly1305_push(&crypto_state,
+						       output + HS_HEADER_LEN,
+						       NULL, hsm_secret->data,
+						       sizeof(hsm_secret->data),
+						       /* Additional data and tag */
+						       NULL, 0, 0))
+		return false;
+
+	return true;
+}
+
+static void destroy_passphrase(char *passphrase)
+{
+	sodium_munlock(passphrase, tal_bytelen(passphrase));
+}
+
+/* Disable terminal echo if needed */
+static bool disable_echo(struct termios *saved_term)
+{
+	if (!isatty(fileno(stdin)))
+		return false;
+
+	if (tcgetattr(fileno(stdin), saved_term) != 0)
+		return false;
+
+	struct termios tmp = *saved_term;
+	tmp.c_lflag &= ~ECHO;
+
+	if (tcsetattr(fileno(stdin), TCSANOW, &tmp) != 0)
+		return false;
+
+	return true;
+}
+
+/* Restore terminal echo if it was disabled */
+static void restore_echo(const struct termios *saved_term)
+{
+	tcsetattr(fileno(stdin), TCSANOW, saved_term);
+}
+
+/* Read line from stdin (uses tal allocation) */
+static char *read_line(const tal_t *ctx)
+{
+	char *line = NULL;
+	size_t size = 0;
+
+	if (getline(&line, &size, stdin) < 0) {
+		free(line);
+		return NULL;
+	}
+
+	/* Strip newline */
+	size_t len = strlen(line);
+	if (len > 0 && line[len - 1] == '\n')
+		line[len - 1] = '\0';
+
+	/* Convert to tal string */
+	char *result = tal_strndup(ctx, line, len);
+	free(line);
+	return result;
+}
+
+const char *read_stdin_pass(const tal_t *ctx, enum hsm_secret_error *err)
+{
+	*err = HSM_SECRET_OK;
+
+	struct termios saved_term;
+	bool echo_disabled = disable_echo(&saved_term);
+	if (isatty(fileno(stdin)) && !echo_disabled) {
+		*err = HSM_SECRET_ERR_TERMINAL;
+		return NULL;
+	}
+
+	char *input = read_line(ctx);
+	if (!input) {
+		if (echo_disabled)
+			restore_echo(&saved_term);
+		*err = HSM_SECRET_ERR_INVALID_FORMAT;
+		return NULL;
+	}
+
+	/* Memory locking is mandatory: failure means we're on an insecure system */
+	if (sodium_mlock(input, tal_bytelen(input)) != 0)
+		abort();
+
+	tal_add_destructor(input, destroy_passphrase);
```

Comment:
Maybe (at end of these!) we could create a generic "mlock_tal_memory()" func:

static void destroy_munlock(const tal_t *t) 
{
    sodium_mlock(t, tal_bytelen(t));
}

void mlock_tal_memory(const tal_t *ptr)
{
    if (sodium_mlock(ptr, tal_bytelen(ptr)) != 0)
		abort();
    tal_add_destructor(ptr, destroy_munlock);
}